linkerd/router/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+use ahash::AHashMap;
+use futures::{future, prelude::*};
+use linkerd_error::Error;
+use linkerd_stack::{layer, NewService, NewSpawnWatch, Oneshot, Param, Service, ServiceExt};
+use std::{
+    fmt::Debug,
+    hash::Hash,
+    marker::PhantomData,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tracing::{debug, error};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RouteKeys<K>(Arc<[K]>)
+where
+    K: Eq + Hash + Clone;
+
+pub trait SelectRoute<Req> {
+    type Key: Eq + Hash + Clone + Debug + Send + Sync + 'static;
+    type Error: std::error::Error + Send + Sync + 'static;
+
+    /// Given a a request, returns the key matching this request.
+    ///
+    /// If no route matches the request, this method returns an error.
+    fn select<'r>(&self, req: &'r Req) -> Result<&Self::Key, Self::Error>;
+}
+
+pub type NewRouteWatch<T, K, L, N> = NewSpawnWatch<T, NewRoute<K, L, N>>;
+
+/// A [`NewService`] that extracts [`RouteKeys`] from the stack target to build a
+/// set of per-route inner services. These services are built first by building
+/// a backend stack for the target and then applying the `L`-typed router layer
+/// to the backend stack.
+#[derive(Clone, Debug)]
+pub struct NewRoute<K, L, N> {
+    route_layer: L,
+    new_backends: N,
+    _marker: PhantomData<fn(K)>,
+}
+
+/// The [`Service`] constructed by [`NewRoute`].
+///
+/// Each request is matched against the route table and routed to the
+/// appropriate
+#[derive(Clone, Debug)]
+pub struct Route<T, K, S> {
+    params: T,
+    routes: Arc<AHashMap<K, S>>,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("unknown route: {0:?}")]
```

Comment:
if the `SelectRoute` selects a key that's not in `self.routes`, that's almost certainly a bug in the `SelectRoute` impl...should this  error message maybe indicate clearly that this is an internal bug in Linkerd?

Reply:
I don't think it's necessarily a 'bug' from the crate's point of view. It doesn't necessarily require that `RouteKeys` is comprehensive for all possible route values. Though, in our specific usage we could have special error handling to that effect.